backends/iptables/iptables.go

```diff
@@ -748,21 +753,15 @@ func (t *iptables) writeDNATRules(svcInfo *serviceInfo, svcName types.Namespaced
 
 		targetPort := t.getTargetPort(svcInfo, endpointPortMap, *epIP)
 
-		// this seems very sly to me. Doing this because there were 2 entries being added
-		// one with the right target port and one with zero
-		// write better logic or verify how baseServiceInfo & endpointPortMap are populated
-		if targetPort == 0 {
-			continue
-		}
 		// DNAT to final destination.
 		args = append(args, "-m", protocol, "-p", protocol, "-j", "DNAT", "--to-destination", net.JoinHostPort(*epIP, strconv.Itoa(targetPort)))
 		t.natRules.Write(args)
 	}
 }
 
-// if the targetPort is string, fetch the value from endpointPortMap
+// if the targetPort is string or portName exists, fetch the value from endpointPortMap
 func (t *iptables) getTargetPort(svcInfo *serviceInfo, endpointPortMap map[string]int32, endpoint string) int {
-	if svcInfo.TargetPortName() != "" {
+	if svcInfo.PortName() != "" || svcInfo.TargetPortName() != "" {
```

Comment:
i wonder if in general , KPNG NEEDS to be dependent on port names for routing rules.  To me it shoudlnt. 

i.e. if

- service 10.1.2.3:8080 points to my hamburger SAAS
- service 10.1.2.3:9090 points to my pizza SAAS

Why cant we blindly just write routing rules to 8080 and 9090.  Who cares what the names are.  After all, the `10.1.2.3:8080` and `10.1.2.3:9090` represent unique primary keys, and are complete in terms of the layer 4 information content required to distinguish these backends.? right?

Reply:
Ah got it.
Kubernetes supports this odd thing where `targetPort=80` and portName=port8080 ... In that case , 
- port name overrides targetPort
but KPNG discarded that name info and then forwareded to the incorrect,  targetPort 80  which was supposed to be replaced w 8080

Reply:
umm isn't what we provide with port overrides in endpoints? I mean, the PortMapping(...) call in api/localv1/endpoint.go

Reply:
@mcluseau previously, the [PortMapping](https://github.com/kubernetes-sigs/kpng/pull/532/files#diff-797a7515d4f869cd70ff1e72a62fab4fd348ad3b509f9a815dbe45ffcbcb3638L34) only worked if `TargetPortName` was specified. This PR added port name check too: `if port.Name != "" ...`.

Reply:
@mikael maybe suggesting there a possible brain fix here ???

Reply:
well, the idea is still to keep k8s business logic outside of the backends. Otherwise, we're just doing that https://xkcd.com/927/ ;-)
As I'm not very available on the project nowadays, I couldn't dig this a lot, but I can see that @mneverov actually did some work in the file I mentioned, so he can have missed it :-) Now, I just wonder why there's a second round of interpretation has, from my "old" memories, the idea of the `PortMappings()` calls was to directly provide the "final form" of argument `enpointPortMap` argument here in `iptables.go`.

Reply:
> `well, the idea is still to keep k8s business logic outside of the backends`

I was not aware of it, got it.

> `the idea of the PortMappings() calls was to directly provide the "final form" of argument enpointPortMap argument here in iptables.go`

From my perspective this PR did not change it.

My interpretation of how port mapping works:

### K8s
Endpoint slice [reconciler](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/endpointslice/reconciler.go#L161) creates endoint slices with port names taken from a corresponding [service port name](https://github.com/kubernetes/kubernetes/blob/fe84992651335ba6dd66e16754e8e416e5144b59/staging/src/k8s.io/endpointslice/utils.go#L87). A Service port [binds](https://github.com/kubernetes/kubernetes/blob/fe84992651335ba6dd66e16754e8e416e5144b59/staging/src/k8s.io/endpointslice/utils.go#L390) `targetName` to the Pod port if the `targetName` is a string (as it is described in [the documentation](https://kubernetes.io/docs/concepts/services-networking/service/#field-spec-ports)). The same [logic](https://github.com/kubernetes/kubernetes/blob/fe84992651335ba6dd66e16754e8e416e5144b59/pkg/api/v1/pod/util.go#L32) is used in the [endpoints controller](https://github.com/kubernetes/kubernetes/blob/fe84992651335ba6dd66e16754e8e416e5144b59/pkg/controller/endpoint/endpoints_controller.go#L443). If a Service port `targetName` is specified as a string and it does not match to any Pod port names then the controllers create an EndpointSlice (or legacy Endpoints) without this port. In case if it is the only port specified in a Service then EndpoinSlice will not have ports at all. If the `targetName` is an integer (i.e. a port 8080), then it is copied to the EndpointSlice port definition.

### KPNG 
KPNGs frontend kube2store watches api-server and [maps](https://github.com/kubernetes-sigs/kpng/blob/6ccd640378e22cba27f119f7ae132026722a5bfc/server/jobs/kube2store/service-event-handler.go#L113) service ports to the internal PortMapping representation. The only difference from k8s is that PortMapping has a separate field `TargetPortName` for the case when the service port is a string.

So, actually [this check](https://github.com/kubernetes-sigs/kpng/pull/532/files#diff-797a7515d4f869cd70ff1e72a62fab4fd348ad3b509f9a815dbe45ffcbcb3638R38) doesn't make sense and should be removed.

For services without selectors the mapping rules are almost the same:
1. If a Service specifies a port name, then it has to match with the EndpointSlice port name. If it doesn't - no routing. One corner case is when Service doesn't specify port name, but EndpointSlice does. In this scenario again, no routing.
2. If a Service specifies `targetPort` (doesn't matter if it is as a string or actual port) it is not part of the decision, i.e. it can be any bogus port and as long as there is a properly defined EndpointSlice, the routing will be created.

```yaml
apiVersion: v1
kind: Service
metadata:
  name: nginx-service
spec:
  ports:
    - port: 80
      protocol: TCP
      targetPort: anything-here

---

apiVersion: discovery.k8s.io/v1
kind: EndpointSlice
metadata:
  name: nginx-service-1
  labels:
    kubernetes.io/service-name: nginx-service
addressType: IPv4
ports:
  - name: ''
    protocol: TCP
    port: 80
endpoints:
  - addresses:
      - "10.244.0.5" # ip of the nginx port
```